Begin implementing deployment logic

alebedev87 opened the pull request:
Handover from https://github.com/openshift/external-dns-operator/pull/18.
Main differences from Steve's PR:
- Missing labelselector/labels added to deployment
- `ExternalDNS.Status` made optional
- Manifests updated to use `openshift-external-dns-operator` namespace

Tried on my local cluster:
```bash
$ oc logs deploy/external-dns-operator -c operator
2021-08-04T12:15:38.200Z	INFO	using operator namespace	{"namespace": "openshift-external-dns-operator"}
2021-08-04T12:15:38.200Z	INFO	using operand namespace	{"namespace": "external-dns"}
2021-08-04T12:15:38.200Z	INFO	using ExternalDNS image	{"image": "docker.io/bitnami/external-dns:latest"}
I0804 12:15:39.252345       1 request.go:668] Waited for 1.03609362s due to client-side throttling, not priority and fairness, request: GET:https://10.217.4.1:443/apis/apps.openshift.io/v1?timeout=32s
2021-08-04T12:15:40.761Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": "127.0.0.1:8080"}
2021-08-04T12:15:43.317Z	INFO	setup	starting externalDNS operator
2021-08-04T12:15:43.320Z	INFO	controller-runtime.manager	starting metrics server	{"path": "/metrics"}
2021-08-04T12:15:43.320Z	INFO	controller-runtime.manager.controller.externaldns_controller	Starting EventSource	{"source": "kind source: /, Kind="}
2021-08-04T12:15:43.320Z	INFO	controller-runtime.manager.controller.externaldns_controller	Starting Controller
2021-08-04T12:15:43.422Z	INFO	controller-runtime.manager.controller.externaldns_controller	Starting workers	{"worker count": 1}
2021-08-04T12:15:43.440Z	INFO	externaldns_controller	created externaldns namespace	{"namespace": "external-dns"}
2021-08-04T12:15:43.492Z	INFO	externaldns_controller	created externalDNS service account	{"namespace": "external-dns", "name": "external-dns-sample"}
2021-08-04T12:15:43.545Z	INFO	externaldns_controller	created externalDNS deployment	{"namespace": "external-dns", "name": "external-dns-sample"}

$ oc get externaldns
NAME     AGE
sample   115s

$ oc -n external-dns get deploy,sa
NAME                                  READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/external-dns-sample   0/1     1            0           73m

NAME                                 SECRETS   AGE
serviceaccount/builder               2         73m
serviceaccount/default               2         73m
serviceaccount/deployer              2         73m
serviceaccount/external-dns-sample   2         73m

```

alebedev87 commented:
/assign @Miciah

candita commented on config/samples/operator_v1alpha1_externaldns.yaml:
```diff
@@ -1,5 +1,15 @@
 metadata:
-  name: externaldns-sample
+  name: sample
+  namespace: external-dns-operator
```
I'll wait for @Miciah to weigh in, but I think you should use `openshift-externaldns-operator` for the namespace

alebedev87 replied:
I was thinking about it too. However OpenShift gave me an error while I tried to create `openshift-external-operator`  namespace for external-dns-opertor and I was cluster-admin. It seems to be that there is some hard coded check for `openshift-*` namespaces.

Miciah replied:
Yeah, the operator should use the `openshift-externaldns-operator` namespace.  What was the error?

alebedev87 replied:
My bad, I naively tested the namespace creation with `oc new-project` which gave me this error:
```bash
$ oc new-project openshift-externaldns-operator
Error from server (Forbidden): project.project.openshift.io "openshift-externaldns-operator" is forbidden: cannot request a project starting with "openshift-"
```
Will change the namespace to `openshift-externaldns-operator`.

candita commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,224 @@
+	if _, err := r.updateExternalDNSDeployment(ctx, desired); err != nil {
+		return true, current, err
+	}
+
```
If there any validation that might return an error here?  If so, return the error to the caller so they can figure out what might be wrong.

candita commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,224 @@
+
+	// update the deployment
+	if _, err := r.updateExternalDNSDeployment(ctx, desired); err != nil {
+		return true, current, err
```
It returns true if there was an error?

alebedev87 replied:
Yes, it was like this before but I'm not particularly against it. This boolean indicates the existence of the deployment, not the error.

candita commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,224 @@
+		for i, serviceType := range externalDNS.Spec.Source.Service.ServiceType {
+			serviceTypeFilter += string(serviceType)
+			// avoid having a trailing comma
+			if len(externalDNS.Spec.Source.Service.ServiceType)-1 > i {
```
You can just do this once after the loop exits.

alebedev87 replied:
Moved the append of the flag outside of `for` loop.

candita replied:
Nit: the len doesn't change, you can just set a variable once outside of the loop.  Or just append the comma every time and remove the last one afterward.

alebedev87 replied:
Done, remove the trailing comma after the loop

candita commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,224 @@
+				serviceTypeFilter += ","
+			}
+			args = append(args, fmt.Sprintf("--service-type-filter=%s", serviceTypeFilter))
+
```
Do you want one instance of the `service-type-filter` parameter, or several?

alebedev87 replied:
It's possible to specify several but I think that it's a little more convenient to see them all on one line.

alebedev87 replied:
Ah, now I see, yeah, it appends every time which doesn't make sense, good catch!

candita commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,224 @@
+	if err := r.client.Create(ctx, depl); err != nil {
+		return fmt.Errorf("failed to create externalDNS deployment %s/%s: %w", depl.Namespace, depl.Name, err)
+	}
+	r.log.Info("created externalDNS deployment", "namespace", depl.Namespace, "name", depl.Name)
```
For Info logs, better to use klog and the flags for verbosity: `klog.V(4).Infof`

alebedev87 replied:
`controller-runtime` based operators and controller use [logr](https://pkg.go.dev/github.com/go-logr/logr) compatible structured logging, not `klog`. About the verbosity level, here were are logging at the default level (without specifying `V()`) so that this will be seen in both [development or production](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.5/pkg/log/zap#Options).

alebedev87 commented:
/assign @candita

alebedev87 commented:
After a discussion with @Miciah, added back the logic for the comparison: current vs desired.

candita commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,261 @@
+	// don't always update (or do simple DeepEqual with) the operand's deployment
+	// as this may result into a "fight" between API/admission webhooks and this controller
+	// example:
+	//  - this contorller creates a deployment with the desired fields A, B, C
```
contorller -> controller

candita commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,261 @@
+		return false, nil
+	}
+
+	if err := r.client.Update(ctx, updated); err != nil {
```
Shouldn't there be some kind of validation on the changes prior to making the update?

alebedev87 replied:
Steve didn't do any validation and for the moment I cannot come up with any either. Also, I think that if something will come we'd better to put it into an admission webhook (validating or mutating for defaults).

candita replied:
We usually do some validation around `Reconcile`, but it could be picked up in a later task.

alebedev87 replied:
Yeah, you are right, these conditions need to be handled and set in the status. I think this can be attacked in the followups.

alebedev87 replied:
Added the handling of the status to TOD in README.md in the [next PR](https://github.com/openshift/external-dns-operator/pull/20).

candita commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,261 @@
+	replicas := int32(1)
+
+	matchLbl := map[string]string{
+		"app.kubernetes.io/name":     "external-dns",
```
Would be nice to have all these strings declared as consts

alebedev87 replied:
I put a constant for externalDNS to `names.go` file in `pkg/operator/controller/` package.
For the label names, I think it's fine to have them defined here. I don't see a big advantage of putting them as `const` at the beginning of the file, I see them more useful to be defined closer to the instantiation of `appsv1.Deployment`.

alebedev87 replied:
Just noticed that in my next PR I used the same value for the toleration's key as the node selector's label.
Will move this value to const and to keep it consistent will do this for all.

candita commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,261 @@
+		},
+	}
+
+	metricsPort := 7979
```
Maybe this should be declared as a const too

candita commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,261 @@
+	}
+
+	metricsPort := 7979
+	for i, zone := range externalDNS.Spec.Zones {
```
Do we have a sense of how many Zones might there be?  Is is possible that there might be port conflicts if we start venturing into the hundreds of Zones?

alebedev87 replied:
I don't think that we have an idea about how many zones can be here, in my mind it's quite a small number (1, 2 digits). I think we may have port conflicts in almost any range. It's the proximity to the popular `8000` range which you think is not good? I didn't see externalDNS listening on `8xxx` ports so far.

Miciah replied:
On OpenShift, there will be 1 or 2 zones, and I doubt any practical use cases outside of OpenShift would call for more than a few zones.

Miciah replied:
We should specify `MaxItems` for `spec.zones` in the CRD though.

alebedev87 replied:
Added `MaxItems` with the value of `10`. @Miciah, let me know which value makes sense here.

alebedev87 replied:
As per discussion in a chat, let's keep it set to 10 for the moment and we'll see if it'll need a change.

candita commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,261 @@
+			updated.Spec.Template.Spec.Containers[0].Image = expected.Spec.Template.Spec.Containers[0].Image
+			changed = true
+		}
+		if !cmp.Equal(current.Spec.Template.Spec.Containers[0].Args, expected.Spec.Template.Spec.Containers[0].Args) {
```
I suppose if there's any chance the args are ordered differently, it might be better to compare them iteratively.

alebedev87 replied:
Well, yes, theoretically you can edit or patch the deployment which would change the order of args. This would result into an update. Don't know about the purpose of such a change though but it can be done accidentally.

alebedev87 replied:
Changed to the comparison of the sorted arg slices

candita commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+)
+
+const (
+	startPortMetrics    = 7979
```
Just for clarity, please rename to metricsPort or metricsStartPort

candita commented on pkg/operator/controller/externaldns/deployment_test.go:
```diff
@@ -0,0 +1,184 @@
+		t.Errorf("expected externalDNS deployment to have %d containers, but found %d", len(externalDNS.Spec.Zones), len(depl.Spec.Template.Spec.Containers))
+	}
+
+	port := 7979
```
7979 -> metricsPort

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+// currentExternalDNSDeployment gets the current externalDNS deployment resource.
+func (r *reconciler) currentExternalDNSDeployment(ctx context.Context, namespace string, externalDNS *operatorv1alpha1.ExternalDNS) (bool, *appsv1.Deployment, error) {
+	depl := &appsv1.Deployment{}
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: namespace, Name: externalDNS.Name}, depl); err != nil {
```
This should use the same name as `desiredExternalDNSDeployment`:
```suggestion
	if err := r.client.Get(ctx, types.NamespacedName{Namespace: namespace, Name: controller.ExternalDNSResourceName(externalDNS)}, depl); err != nil {
```

Miciah commented on config/default/manager_auth_proxy_patch.yaml:
```diff
@@ -22,6 +22,3 @@ spec:
           requests:
             cpu: 100m
             memory: 20Mi
-      - name: operator
-        args:
-        - "--metrics-bind-address=127.0.0.1:8080"
```
Why delete this?

alebedev87 replied:
I thought it was added to add the patch to find the context of where auth proxy container has to be added. But it works without it. Also `metrics-bind-address` is set to `127.0.0.1:8080` by default.

Miciah commented on config/samples/operator_v1alpha1_externaldns.yaml:
```diff
@@ -1,5 +1,15 @@
+        namespace: openshift-external-dns-operator
+  source:
+    type: Service
+  zones: ["dnszone"]
```
Would it be better to leave `spec.zones` unspecified?  On platforms where zones can be autodetected, leaving it unspecified will allow this sample to work out of the box, if the credentials can be configured appropriately.  On other platforms, the sample won't work anyway.

alebedev87 replied:
What I had in my mind is more like having the sample which encourages the user to think about where the records should be published. The reason for this is when I first managed to get all the credentials setup ExternalDNS started to try to publish DNS records for all the services in the whole cluster and I was rather thankful of having the dns zone which didn't match any real zone ID. Once the user feels more comfortable with how ExternalDNS works, the filters can be removed.
I actually added `namespace` and `annotationFilter` to the sample too, for the same reason.

Miciah replied:
That makes sense.  The README suggested running `oc apply -k config/samples`, so I figured the sample was supposed to be ready to use out of the box, but it makes sense not to have the out-of-the-box example automatically do _too_ much.

Miciah commented on pkg/operator/config/config.go:
```diff
@@ -17,10 +17,11 @@ limitations under the License.
-	DefaultMetricsAddr       = "127.0.0.1:8080"
-	DefaultOperatorNamespace = "externaldns-operator"
-	DefaultOperandNamespace  = "externaldns"
+	DefaultExternalDNSImage     = "docker.io/bitnami/external-dns:latest"
```
We should use a quay.io pullspec, but maybe that should be in a follow-up.

alebedev87 replied:
I'd like to do it too but I put on hold the setup of the mirroring for ART built external-dns images.

Miciah commented on pkg/operator/config/config.go:
```diff
@@ -17,10 +17,11 @@ limitations under the License.
-	DefaultOperandNamespace  = "externaldns"
+	DefaultExternalDNSImage     = "docker.io/bitnami/external-dns:latest"
+	DefaultMetricsAddr          = "127.0.0.1:8080"
+	DefaultOperatorNamespace    = "openshift-external-dns-operator"
+	DefaultOperandNamespace     = "external-dns"
```
The goal being to make this upstreamable, would it make more sense for the default namespace not to include "openshift-", and to include a set of manifests or patches to configure the operator and operand to use "openshift-"-prefixed namespaces?

alebedev87 replied:
Hm, that sounds familiar. https://github.com/openshift/external-dns-operator/pull/19#discussion_r682107340

Miciah replied:
Oops!  Hm, I guess we'll need to de-OpenShift-ize some other things anyway, so I guess using "openshift-" is fine.  However, if the operator namespace has the "openshift-" prefix, so should the operand namespace.

alebedev87 replied:
I'm fine to revert all back to `external-dns-operator` namespace.

alebedev87 replied:
It's now `external-dns-operator`, will have to do something for OpenShift intergrations in the future.

Miciah commented on pkg/operator/controller/externaldns/controller.go:
```diff
@@ -77,6 +81,7 @@ func New(mgr manager.Manager, cfg Config) (controller.Controller, error) {
 // the object match the desired state.
 func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	result := reconcile.Result{}
+	requeueResult := reconcile.Result{RequeueAfter: time.Duration(r.config.RequeuePeriodSeconds) * time.Second, Requeue: true}
```
Specifying `Requeue` is superfluous if you specify a non-zero value for `RequeueAfter`.

Miciah commented on pkg/operator/controller/externaldns/controller.go:
```diff
@@ -88,16 +93,16 @@ func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
-		// Return if the externalDNS namespace cannot be created since
-		// resource creation in a namespace that does not exist will fail.
-		return result, fmt.Errorf("failed to ensure externalDNS namespace: %v", err)
+		return requeueResult, fmt.Errorf("failed to ensure externalDNS namespace: %w", err)
```
Specifying a non-zero value result is superfluous when the error value is non-nil.  Specifying `Requeue` or `RequeueAfter` is only really needed if there was no error but you want to terminate reconciliation and start it over from the beginning.

alebedev87 replied:
I thought of not re-queuing right after the failure (of any secondary resource creation). Some errors would not change after waiting for some time or not but some might disappear. Also, if it's not a fixable problem logging an error with some re-queue time in between would not pollute the logs that fast.

Miciah replied:
I mean, you're returning `fmt.Errorf(...)`, which is a non-nil error value, and controller-runtime will requeue the reconciliation request immediately anyway when it gets that non-nil error value, so using `return result, fmt.Errorf(...)` should have the same effect as using `return requeueResult, fmt.Errorf(...)`.

alebedev87 replied:
It's a pity it's not possible to specify `RequeueAfter` for error cases. I removed all requeue logic as for the moment it doesn't make much sense for the success case.

Miciah commented on pkg/operator/controller/externaldns/controller.go:
```diff
@@ -88,16 +93,16 @@ func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
+	_, sa, err := r.ensureExternalDNSServiceAccount(ctx, r.config.Namespace, externalDNS)
 	if err != nil {
-		return result, fmt.Errorf("failed to ensure externalDNS service account: %w", err)
-	} else if !haveServiceAccount {
-		return result, fmt.Errorf("failed to get externalDNS service account: %w", err)
```
Seems like it makes sense to keep this check.

alebedev87 replied:
I don't know, just to return a different error message? `failed to get` vs `failed to ensure`?

Miciah replied:
If the serviceaccount isn't there, `sa` will be nil, and `desiredExternalDNSDeployment` will deference a nil pointer.  (Granted, it's extremely unlikely that `ensureExternalDNSServiceAccount` will return false and a nil error value, but it could happen if it does a `Create` and a subsequent `Get` and the `Get` gets a non-found response.  Anyway, we get a Boolean return value here, so it makes sense to check it.)

alebedev87 replied:
Yes, you are right, good catch, thank you!

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+	metricsStartPort    = 7979
+	appNameLabel        = "app.kubernetes.io/name"
+	appInstanceLabel    = "app.kubernetes.io/instance"
+	masterNodeRoleLabel = "node-role.kubernetes.io/master"
```
This label is deprecated in Kubernetes 1.20:
```suggestion
	controlPlaneNodeRoleLabel = "node-role.kubernetes.io/control-plane"
```

Miciah replied:
That's strange.  I'll try to find out why that label isn't there.

alebedev87 replied:
It's a gap on OCP side, [BZ filed](https://bugzilla.redhat.com/show_bug.cgi?id=1995858).

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+}
+
+// ensureExternalDNSDeployment ensures that the externalDNS deployment exists.
+// Returns a boolean if the deployment exists, and an error when relevant.
```
```suggestion
// Returns a Boolean value indicating whether the deployment exists, a pointer to the deployment, and an error when relevant.
```

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+		if err := r.createExternalDNSDeployment(ctx, desired); err != nil {
+			return false, nil, err
+		}
+		// getting the deployment from API in case a third party updated it after the creation
```
The reason for doing another get here is so that we have the object with fields that the API fills in, such as `resourceVersion` and defaulted fields, without which a subsequent update would fail.

alebedev87 replied:
Webhooks can modify the fields too, that's why I wrote it less precise. But it's kinda misleading, I agree.
Let me rephrase it.

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+}
+
+// WIP function for generating container specs for one container at a time.
+func populateExternalDNSContainer(image string, metricsPort int, zone string, externalDNS *operatorv1alpha1.ExternalDNS) (*corev1.Container, error) {
```
Is this still WIP?

Would it make sense to return a non-pointer value?

Tiny suggestion: Would it make sense to name this `buildExternalDNSContainer`, to be consistent with the terminology in the "failed to build" error message in `ensureExternalDNSDeployment`?

alebedev87 replied:
Well, yes, it still needs the credentials logic.
Renamed to `buildExternalDNSContainer`.
Returning a value would make sense to not deference the pointer when it's added to `Spec.Containers`. Don't know whether something like return value optimization exists in Golang, if it does then it doesn't make much sense to return by pointer indeed.

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+
+// WIP function for generating container specs for one container at a time.
+func populateExternalDNSContainer(image string, metricsPort int, zone string, externalDNS *operatorv1alpha1.ExternalDNS) (*corev1.Container, error) {
+	name := "externaldns-" + zone
```
Is `zone` necessarily valid in a container name?  It might be better to hash the name, or just use a number ("externaldns-1", "externaldns-2", ...).

alebedev87 replied:
Replaced the zone with a sequence number.

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+	}
+	args = append(args, fmt.Sprintf("--provider=%s", provider))
+
+	//TODO: Add provider credentials logic
```
The caller could look up credentials.

alebedev87 replied:
Not sure I understood what you meant.
I saw this in the enhancement doc:
> On OpenShift, the ExternalDNS Operator could create a credentialsrequest.cloudcredential.openshift.io resource so that the cluster administrator would not have to worry about providing cloud credentials by hand when creating an ExternalDNS operator resource. The ExternalDNS Operator would the be responsible for configuring the relevant ExternalDNS deployments to consume the secret produced from the credentialsrequest.cloudcredential.openshift.io resource.

But I thought this would be an option for OpenShift deployments, for the main (Kubernetes) case I thought we would still wait for the creator of `externaldns` resource to provide a secret.

Miciah replied:
I mean, the "TODO" could be in `desiredExternalDNSDeployment`, and once `desiredExternalDNSDeployment` had logic to look up the credentials, it could pass them to `buildExternalDNSContainer`.

alebedev87 replied:
Yes, I can add a TODO for the retrieval of the credentials secret into `desiredExternalDNSDeployment`. Unfortunately it's the only common thing between them, how the credentials should be consumed in externalDNS POD  is specific to each provider. So, `buildExternalDNSContainer` would still need to manage these differences therefore still needs a TODO.

alebedev87 replied:
I think that how and where we lookup the credentials can be revisited when we get closer to the implementation, in the followup PRs. And I'll keep in mind this remark.

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+
+	//TODO: Add provider credentials logic
+
+	source, ok := sourceStringTable[externalDNS.Spec.Source.Type]
```
The caller could do the source lookup, same way and with the same benefits that I described in my comment about the provider lookup.

alebedev87 replied:
You mean you'd prefer to see something like:
```golang
func buildExternalDNSContainer(image, zone, provider string, source *ExternalDNSSource)
```
as a declaration instead of passing the whole `ExternalDNS` instance?

Miciah replied:
I'm just suggesting that the `sourceStringTable[externalDNS.Spec.Source.Type]` lookup could be in `desiredExternalDNSDeployment`, and then `buildExternalDNSContainer` couldn't get an error.  If `desiredExternalDNSDeployment` did both the provider and source map lookups, then `buildExternalDNSContainer` would never return a non-nil error value, and the error return value could be deleted.  (Maybe as more options are added, more reasons to return an error will arise though?)

alebedev87 replied:
Ah ok, I see. As a matter of fact the provider mapping can be replaced with a simple `ToLower()` as the values are already validated by OpenAPI of CRD and their ExternalDNS's equivalents are the same but in lower case.
The source unfortunately has 1 option which is drastically different in API and ExternalDNS: `OpenShiftRoute` vs `openshift-route`. The rest is also just a different case.
Well, ok, I'll pass them both as params, this will make the function a pure builder with no error handling, we'll see if this will work out (number of input params vs error output param).

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+
+	source, ok := sourceStringTable[externalDNS.Spec.Source.Type]
+	if !ok {
+		return nil, fmt.Errorf("%s is not a supported source type", externalDNS.Spec.Source.Type)
```
```suggestion
		return nil, fmt.Errorf("unsupported source type: %q", externalDNS.Spec.Source.Type)
```

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+		}
+
+		// avoid having a trailing comma
+		args = append(args, fmt.Sprintf("--service-type-filter=%s", serviceTypeFilter[0:len(serviceTypeFilter)-1]))
+
+		if publishInternal {
+			args = append(args, "--publish-internal-services")
```
Any idea why `--publish-internal-services` is a separate flag from `--service-type-filter`?  Is it a legacy flag, or is publishing ClusterIP services dangerous or something such that it calls for a separate flag?

alebedev87 replied:
They preferred this to be an opt-in flag: https://github.com/kubernetes-sigs/external-dns/pull/278

Miciah replied:
So it looks like it's a flag to say, "Yes, I really did mean to include ClusterIP services."  Do we need the same kind of "Yes, I really mean it" flag in the CRD?  I'd rather not, but the fact that upstream added the flag makes me wary.

alebedev87 replied:
No, didn't see anything like this for CRD. There are some other guard flags for headless services but that not quite the same case as for `ClusterIP` as headless is not a service type so you cannot set in `--service-type-filter`.
```bash
  --publish-host-ip              Allow external-dns to publish host-ip for 
                                 headless services (optional)
  --always-publish-not-ready-addresses  
                                 Always publish also not ready addresses for 
                                 headless services (optional)
```

alebedev87 replied:
Looked at this thing one more time. Actually you were right about the legacy.

At the time when `--publish-internal-services` was added there was no `--service-type-filter` flag. There was no way to fine tune deeper than the source level, so you could only say: I want services as source (`--source=service`) and yes I want internal ClusterIPs to be published too (`--publish-internal-services`).

Then they added  `--service-type-filter` flag in [PR#653](https://github.com/kubernetes-sigs/external-dns/pull/653) which defaulted to `all` which kept the old combination of `--source=service --publish-internal-services` still relevant. However if the filter is set to `ClusterIP` you have to specify the unnecessary `--publish-internal-services`, otherwise the IP [will not be found](https://github.com/kubernetes-sigs/external-dns/blob/master/source/service.go#L461).

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+	// as this may result into a "fight" between API/admission webhooks and this controller
+	// example:
+	//  - this controller creates a deployment with the desired fields A, B, C
+	//  - API adds some default fields D, E, F (i.e. metadata, imagePullPullPolicy, dnsPolicy)
```
```suggestion
	//  - API adds some default fields D, E, F (e.g. metadata, imagePullPullPolicy, dnsPolicy)
```

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+	return true, nil
+}
+
+// externalDNSDeploymentChange evaluates whether or not a deployment update is necessary.
```
```suggestion
// externalDNSDeploymentChanged evaluates whether or not a deployment update is necessary.
```

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+	changed := false
+	updated := current.DeepCopy()
+
+	if len(current.Spec.Template.Spec.Containers) > 0 && len(expected.Spec.Template.Spec.Containers) > 0 {
+		if current.Spec.Template.Spec.Containers[0].Image != expected.Spec.Template.Spec.Containers[0].Image {
+			updated.Spec.Template.Spec.Containers[0].Image = expected.Spec.Template.Spec.Containers[0].Image
+			changed = true
+		}
+		currArgs := append([]string{}, current.Spec.Template.Spec.Containers[0].Args...)
+		expArgs := append([]string{}, expected.Spec.Template.Spec.Containers[0].Args...)
+		sort.Strings(currArgs)
+		sort.Strings(expArgs)
+		if !cmp.Equal(currArgs, expArgs) {
+			updated.Spec.Template.Spec.Containers[0].Args = expected.Spec.Template.Spec.Containers[0].Args
+			changed = true
+		}
+	}
```
We want to be as impermissive of changes to operator-managed resources as possible—stomp any and all changes by users, but allow for defaulting behavior by the API and mutating admission plugins.  We can tighten the update logic up in a follow-up, but once people start using the operator, they may start doing all sorts of weird and unanticipated hacks if we allow it, and then it becomes much more difficult to tighten the operator's control over its resources.

alebedev87 replied:
Do you have in mind some concrete fields we can additionally check?

alebedev87 replied:
Updated the change detection logic and unit tests. Hashing of the zone is added too (needed for the change detecction).
Put everything in a separate commit, just in case: https://github.com/openshift/external-dns-operator/pull/19/commits/c0c6c2b55751ad45c310a3872a138ad941cc8aa6

alebedev87 replied:
As discussed in the chat, the last implementation from https://github.com/openshift/external-dns-operator/commit/7ac3c115f2696eb94d4ed5f1298374a19eb6c20d looks fine to begin with.

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+		}
+	}
+
+	if !changed {
+		updated = nil
+	}
```
This seems unnecessary—once the function returns, `updated` goes away anyway.

Miciah commented on cmd/external-dns-operator/main.go:
```diff
@@ -41,6 +41,7 @@ func main() {
 	flag.StringVar(&opCfg.OperatorNamespace, "operator-namespace", operatorconfig.DefaultOperatorNamespace, "The namespace that the operator is running in.")
 	flag.StringVar(&opCfg.OperandNamespace, "operand-namespace", operatorconfig.DefaultOperandNamespace, "The namespace that ExternalDNS containers should run in.")
 	flag.StringVar(&opCfg.ExternalDNSImage, "externaldns-image", operatorconfig.DefaultExternalDNSImage, "The container image used for running ExternalDNS.")
+	flag.IntVar(&opCfg.RequeuePeriodSeconds, "requeue-period", operatorconfig.DefaultRequeuePeriodSeconds, "How much time to wait after a failed reconcile (in seconds).")
```
What's the use case for setting a different requeue period?

alebedev87 replied:
Removed the requeue logic as it doesn't make much sense for the success case, at least so far.

Miciah commented on pkg/operator/controller/externaldns/deployment.go:
```diff
@@ -0,0 +1,274 @@
+		return false, nil, fmt.Errorf("failed to build externalDNS deployment: %w", err)
+	}
+
+	exist, current, err := r.currentExternalDNSDeployment(ctx, namespace, externalDNS)
```
Instead of passing `namespace` and `externalDNS` to `currentExternalDNSDeployment` and other functions which then need to construct the namespaced name, it would be simpler to compute a `types.NamespacedName` value at the top of `ensureExternalDNSDeployment` and pass it down and use it throughout this function and its callees.  (BTW, a `types.NamespacedName` value is stringified as `namespace/name`, so it is convenient to use in error messages and log messages as well.)

alebedev87 replied:
Done, for service account and namespace files too

Miciah reviewed (commented):
`TestExternalDNSDeploymentChanged` should have test cases for zone added, changed, or deleted.

candita commented:
> `TestExternalDNSDeploymentChanged` should have test cases for zone added, changed, or deleted.

I think this is one comment that was overlooked.  Otherwise, lgtm.

alebedev87 commented:
@candita: I addressed the comment about the zones in `deployment_test.go`. It's just that the new test cases don't mention the zones but containers, as zone == container in terms of the externalDNS deployment. The test cases are:
- `if externalDNS misses container` when a new zone was added
- `if externalDNS has extra container` when a zone was deleted

The other test cases cover the update of a zone: image or args (the only 2 things we care about for the moment).

candita commented:
/lgtm
